python/lsst/ctrl/pool/parallel.py

```diff
@@ -473,3 +477,68 @@ def parseAndRun(cls, *args, **kwargs):
         super(BatchPoolTask, cls).parseAndRun(*args, **kwargs)
         pool.exit()
 
+
+class BatchTaskRunner(TaskRunner):
+    """Run a Task individually on a list of inputs using the MPI process pool"""
+    def __init__(self, *args, **kwargs):
+        TaskRunner.__init__(self, *args, **kwargs)
+        if self.numProcesses > 1:
+            self.log.warn("Multiprocessing arguments (-j %d) ignored since using batch processing" %
+                          self.numProcesses)
+            self.numProcesses = 1
+
+    def run(self, parsedCmd):
+        """Run the task on all targets"""
+        resultList = None
+
+        import multiprocessing
+        self.prepareForMultiProcessing()
+        pool = Pool()
+
+        if self.precall(parsedCmd):
+            targetList = self.getTargetList(parsedCmd)
+            if len(targetList) > 0:
+                parsedCmd.log.info("Processing %d targets with a pool of %d processes..." %
+                                   (len(targetList), pool.size))
+                # Run the task using self.__call__
+                resultList = pool.map(self, targetList)
+            else:
+                log.warn("Not running the task because there is no data to process; "
+                    "you may preview data using \"--show data\"")
+                resultList = []
+
+        return resultList
+
+    @abortOnError
+    def __call__(self, cache, args):
+        """Run the Task on a single target
+
+        Strips out the process pool 'cache' argument.
+
+        Brings down the entire job if an exception is not caught (i.e., --doraise).
+        """
+        return TaskRunner.__call__(self, args)
+
+
+class BatchParallelTask(BatchCmdLineTask):
+    """Runs the BatchCmdLineTask in parallel
+
+    Use this subclass of BatchCmdLineTask if you don't need to use the Pool
+    directly, but just want to iterate over many objects (like a multi-node
+    version of the '-j' command-line argument).
+    """
+    RunnerClass = BatchTaskRunner
+
+    @classmethod
+    def _makeArgumentParser(cls, *args, **kwargs):
+        kwargs.pop("doBatch", False)
+        kwargs.pop("add_help", False)
+        return super(BatchCmdLineTask, cls)._makeArgumentParser(*args, **kwargs)
+
+    @classmethod
+    def parseAndRun(cls, *args, **kwargs):
+        """Parse an argument list and run the command"""
+        pool = startPool()  # Start the pool early, to minimise the amount of duplicated work
+        results = super(BatchParallelTask, cls).parseAndRun(*args, **kwargs)
```

Comment:
I tried very hard not to use it!  But I found that I *must* use it if I want to call the base class implementation of one method that calls another method I've implemented in my subclass.  Consider the following simplified case:

    class Foo(object):
        @staticmethod
        def name():
            print "Foo!"
        @classmethod
        def run(cls):
            cls.name()
    
    class Bar(Foo):
        @staticmethod
        def name():
            print "Bar!"
        @classmethod
        def run(cls):
            Foo.run()

I have a base class (`Foo`) and a subclass (`Bar`).  I want to call `Bar.run()` which defers the implementation to the base class (and presumably does something different with the result, but that's left out of this simplified case), but that implementation calls the `name` method on my object.  If I implement `Bar.run` as I have above, without using `super` and calling `Foo.run`, I get:

    >>> Bar.run()
    Foo!

`Foo.run` gets `cls = Foo`, and so calls `Foo.name()`.  If I want `Foo.run` to call `Bar.name`, then I have to use `super` so I get an object that is still a `Bar` while I'm calling `Foo.run`.  With this implementation:

    class Bar(Foo):
        @staticmethod
        def name():
            print "Bar!"
        @classmethod
        def run(cls):
            super(Bar, cls).run()

then I get the desired result:

    >>> Bar.run()
    Bar!

In my case, I want `BatchParallelTask.parseAndRun` to call the base class `BatchCmdLineTask.parseAndRun()`, but that calls the `_makeArgumentParser` method and I want the implementation in the subclass to fire, not the one in the base class.  So, as far as I can tell, I absolutely *must* use `super`.